### PR TITLE
Wrapped machine

### PR DESCRIPTION
**Pull request by Christian-B**
All PRs must be done at the same time.

https://github.com/SpiNNakerManchester/PACMAN/pull/222
https://github.com/SpiNNakerManchester/SpiNNMan/pull/161
https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/435
https://github.com/SpiNNakerManchester/sPyNNaker/pull/656

tested by 
https://github.com/SpiNNakerManchester/sPyNNaker8/pull/218
But that adds no new code

### CODE REVIEW

**Comment by dkfellows** [comment]
Do we have tests for checking if we can go from the chip at `5,0` (on a 12&times;12 machine, wrapped) to its board-local coordinates and back?

**Comment by Christian-B** [comment]
> Do we have tests for checking if we can go from the chip at `5,0` (on a 12×12 machine, wrapped) to its board-local coordinates and back?

test_virtual_machine.py methods test_fullwrap, test_vertical_wrap, test_horizontal_wrap and test_no_wrap test all chips on a 3 board setup to local and back.
the famous 5.0 did go wrong first until I remembered that the M in BOMDAS also means mod.

File: spinn_machine/machine.py [hunk 1]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+        if origin is None:
+            self._origin = ""
+        else:
+            self._origin = origin
+
+    @abstractmethod
+    def multiple_48_chip_boards(self):
+        """
+        Checks that the width and height coorespond to the expected size for a
+        multipy board machine made up of more than one 48 chip board.
```
Comment (dkfellows) [inline spinn_machine/machine.py]: Speel chuck: `multipy board` &rarr; `multi-board`

File: spinn_machine/machine.py [hunk 2]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+
+        The assumption is that any size machine can be supported but that none
```
Comment (dkfellows) [inline spinn_machine/machine.py]: “but that none”?

File: spinn_machine/machine.py [hunk 3]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+
+        +++++
+        return: True if this machine can have multiple 48 chip boards
+        """
+
+    @abstractmethod
+    def get_xys_by_ethernet(self, ethernet_x, ethernet_y):
+        """
+        Yields the protential x,y locations of all the chips on the board #
```
Comment (dkfellows) [inline spinn_machine/machine.py]: `#` &rarr; `\`

File: spinn_machine/machine.py [hunk 4]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+
+        Warning: GIGO! This methods assumes that ethernet_x and ethernet_y are
+        the local 0,0 of an existing board,
+        within the width and height of the machine.
+
+        :param ethernet_x: The x coordinate of a (local 0,0) legal ethernet
+        chip
+        :param ethernet_y: The x coordinate of a (local 0,0) legal ethernet
+        chip
+        :return: Yeilds the (x, Y) coordinated of all the protential chips on
```
Comment (dkfellows) [inline spinn_machine/machine.py]: Speelchuck: `Yeilds` &rarr; `Yields`

File: spinn_machine/machine.py [hunk 5]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+         this board.
```
Comment (dkfellows) [inline spinn_machine/machine.py]: Indent error

File: spinn_machine/machine.py [hunk 6]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+
+        It is the callers responsibility to check the validity of this call
+        before making it or the validty of the result.
+
+        Warning: GIGO! This methods assumes that x and y are within the
+        width and height of the machine, and that the link goes to another
+        chip on the machine.
+
+        On machine without full wraparound it is possible that this method
+        generates x_Y values that fall outside of the legal values including
```
Comment (dkfellows) [inline spinn_machine/machine.py]: `x_Y` &rarr; `x,y`?

File: spinn_machine/machine.py [hunk 7]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+        negatives values, x = width or y = height.
```
Comment (dkfellows) [inline spinn_machine/machine.py]: `negatives`&rarr;`negative`

File: spinn_machine/machine.py [hunk 8]
```diff
@@ -89,16 +121,213 @@ def __init__(self, chips, boot_x, boot_y):
...
+    def get_global_xy(self, local_x, local_y, ethernet_x, ethernet_y):
+        """
+        Converts the local x and y cooridinates into global x,y coordinates,
+        under the assumption that they are on the board with local 0,0 at
+        ethernet_x, ethernet_y
+
+        This method does take wrap arounds into consideration.
+
+        GIGO: This method does not check if input parameters make sense,
+        nor does it check if there is a chip at the resulting global xy
```
Comment (dkfellows) [inline spinn_machine/machine.py]: `xy`&rarr;`x,y`

File: spinn_machine/machine_factory.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+from spinn_machine.no_wrap_machine import NoWrapMachine
+from spinn_machine.horizontal_wrap_machine import HorizontalWrapMachine
+from spinn_machine.vertical_wrap_machine import VerticalWrapMachine
+from spinn_machine.full_wrap_machine import FullWrapMachine
+
+
+def machine_from_size(width, height, chips=None, origin=None):
+    """
+    Create a machine with the assumed wrap arround based on the sizes.
```
Comment (dkfellows) [inline spinn_machine/machine_factory.py]: `arround`&rarr;`around`

File: spinn_machine/machine_factory.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    This could include a machine with no wrap arrounds, only vertical ones,
```
Comment (dkfellows) [inline spinn_machine/machine_factory.py]: `arround`&rarr;`around`

File: spinn_machine/machine_factory.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    only horizontal ones or both.
+
+    Note: If the sizes do not match the ones for a known wrap arround machine,
```
Comment (dkfellows) [inline spinn_machine/machine_factory.py]: `arround`&rarr;`around`

File: spinn_machine/machine_factory.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    else:
+        if height % 12 == 0:
+            return VerticalWrapMachine(width, height, chips, origin)
+        else:
+            return NoWrapMachine(width, height, chips, origin)
+
+
+def machine_from_chips(chips):
+    """
+    Create a machine with the assumed wrap arround based on the sizes.
```
Comment (dkfellows) [inline spinn_machine/machine_factory.py]: `arround`&rarr;`around`

File: spinn_machine/no_wrap_machine.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
+from .machine import Machine
+from spinn_utilities.overrides import overrides
+
+
+class NoWrapMachine(Machine):
+    def __init__(self, width, height, chips=None, origin=None):
+        """
+        Creates an unwrapped machine
+
+        :param width: The width of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/no_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/no_wrap_machine.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        :param height: The height of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/no_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/vertical_wrap_machine.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
+from spinn_utilities.overrides import overrides
+from .machine import Machine
+
+
+class VerticalWrapMachine(Machine):
+    def __init__(self, width, height, chips=None, origin=None):
+        """
+        Creates a vertically wrapped machine
+
+        :param width: The width of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/vertical_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/vertical_wrap_machine.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        :param height: The height of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/vertical_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/full_wrap_machine.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+class FullWrapMachine(Machine):
+    """
+    Provides an extension of Machine which supports full wrap arounds
+    """
+
+    def __init__(self, width, height, chips=None, origin=None):
+        """
+        Creates a fully wrapped machine
+
+        :param width: The width of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/full_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/full_wrap_machine.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        :param height: The height of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/full_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/horizontal_wrap_machine.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
+from spinn_utilities.overrides import overrides
+from .machine import Machine
+
+
+class HorizontalWrapMachine(Machine):
+    def __init__(self, width, height, chips=None, origin=None):
+        """
+        Creates a horizontally wrapped machine.
+
+        :param width: The width of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/horizontal_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/horizontal_wrap_machine.py [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+        :param height: The height of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/horizontal_wrap_machine.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/machine_factory.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+    This could include a machine with no wrap arrounds, only vertical ones,
+    only horizontal ones or both.
+
+    Note: If the sizes do not match the ones for a known wrap arround machine,
+    no wrap arrounds is assumed.
+
+    :param width: The width of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/machine_factory.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/machine_factory.py [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    :param height: The height of the machine excluding any vertical chips
```
Comment (dkfellows) [inline spinn_machine/machine_factory.py]: `vertical`&rarr;`virtual`?

File: spinn_machine/virtual_machine.py [hunk 23]
```diff
@@ -154,130 +238,33 @@ def __init__(
...
-        for chip in self._unreachable_outgoing_chips:
-            configured_chips.remove(chip)
-        for chip in self._unreachable_incoming_chips:
-            configured_chips.remove(chip)
+        for (eth_x, eth_y) in ethernet_chips:
+            for x_y in self._machine.get_xys_by_ethernet(eth_x, eth_y):
+                if x_y not in down_chips:
+                    configured_chips[x_y] = (eth_x, eth_y)
+
+        # TODO This needs to change as it previous checked against empty
```
Comment (dkfellows) [inline spinn_machine/virtual_machine.py]: I see a `TODO`…

File: spinn_machine/horizontal_wrap_machine.py (hunk 19 above)
Comment (Christian-B) [reply spinn_machine/horizontal_wrap_machine.py]: fix everywhere

**Pull request merged**